data-prepper-plugins/parquet-codecs/src/main/java/org/opensearch/dataprepper/plugins/codec/parquet/ParquetOutputCodec.java

```diff
@@ -4,48 +4,248 @@
+
+    private static final String TIME_PATTERN_REGULAR_EXPRESSION = "\\%\\{.*?\\}";
+    private static final Pattern SIMPLE_DURATION_PATTERN = Pattern.compile(TIME_PATTERN_REGULAR_EXPRESSION);
+    private String key;
+    private final String bucket;
+    private final String region;
+    private static final List<String> primitiveTypes = Arrays.asList("int", "long", "string", "float", "double", "bytes");
+
 
     @DataPrepperPluginConstructor
     public ParquetOutputCodec(final ParquetOutputCodecConfig config) {
-        // TODO: implement initiate config
+        Objects.requireNonNull(config);
+        this.config = config;
+        this.region = config.getRegion();
+        this.bucket = config.getBucket();
     }
 
-
     @Override
-    public void start(final OutputStream outputStream) throws IOException {
-        // TODO: do the initial wrapping
+    public synchronized void start(final OutputStream outputStream) throws IOException {
+        this.s3Client = buildS3Client();
+        buildSchemaAndKey();
+        final S3OutputFile s3OutputFile = new S3OutputFile(s3Client, bucket, key);
+        buildWriter(s3OutputFile);
+    }
+
+    public synchronized void start(File file) throws IOException {
+        LocalOutputFile localOutputFile =new LocalOutputFile(file);
+        buildSchemaAndKey();
+        buildWriter(localOutputFile);
+    }
+
+    void buildSchemaAndKey() throws IOException {
+        if (config.getSchema() != null) {
+            schema = parseSchema(config.getSchema());
+        } else if(config.getFileLocation()!=null){
+            schema = ParquetSchemaParser.parseSchemaFromJsonFile(config.getFileLocation());
+        }else if(config.getSchemaRegistryUrl()!=null){
+            schema = parseSchema(ParquetSchemaParserFromSchemaRegistry.getSchemaType(config.getSchemaRegistryUrl()));
+        }else if(checkS3SchemaValidity()){
+            schema = ParquetSchemaParserFromS3.parseSchema(config);
+        }
+        else{
+            LOG.error("Schema not provided.");
+            throw new IOException("Can't proceed without Schema.");
+        }
+        key = generateKey();
+    }
+
+    private void buildWriter(OutputFile outputFile) throws IOException {
+        writer = AvroParquetWriter.<GenericRecord>builder(outputFile)
+                .withSchema(schema)
+                .build();
     }
 
     @Override
-    public void complete(final OutputStream outputStream) throws IOException {
-        // TODO: Close the output stream
+    public void writeEvent(final Event event, final OutputStream outputStream,final String tagsTargetKey) throws IOException {
+        final GenericData.Record parquetRecord = new GenericData.Record(schema);
+        final Event modifiedEvent;
+        if (tagsTargetKey != null) {
+            modifiedEvent = addTagsToEvent(event, tagsTargetKey);
+        } else {
+            modifiedEvent = event;
+        }
+        for (final String key : modifiedEvent.toMap().keySet()) {
```

Comment:
Since you are anyways creating a new parquetRecord, I think it makes sense to add the "tagsTargetKey" also here without creating a 'modifiedEvent', right? I think you should be write like

```
      if (tagsTargetKey != null) {
            parquetRecord.put(tagsTargetKey, event.getMetadata().getTags());
       }. 
       writer.write(parquestRecord)

Reply:
Hi Krishna, the logic behing modifying the event itself is because we are generating schema from event in case user doesn't supply any schema.If event won't be modified to contain tags target key and tags then parquet record won't be able to store it either.